src/prefect/server/api/concurrency_limits.py

```diff
@@ -33,20 +50,59 @@ async def create_concurrency_limit(
 
     For more information, see https://docs.prefect.io/v3/develop/task-run-limits.
     """
-    # hydrate the input model into a full model
-    concurrency_limit_model = schemas.core.ConcurrencyLimit(
-        **concurrency_limit.model_dump()
-    )
+    # Always use V2 for create, return V1 shape
+    v2_name = f"tag:{concurrency_limit.tag}"
 
     async with db.session_context(begin_transaction=True) as session:
-        model = await models.concurrency_limits.create_concurrency_limit(
-            session=session, concurrency_limit=concurrency_limit_model
+        # Check if exists for upsert behavior
+        existing = await cl_v2_models.read_concurrency_limit(
+            session=session, name=v2_name
         )
 
-    if model.created >= now("UTC"):
-        response.status_code = status.HTTP_201_CREATED
+        if existing:
+            # Update existing (V1 upsert behavior)
+            await cl_v2_models.update_concurrency_limit(
+                session=session,
+                concurrency_limit_id=existing.id,
+                concurrency_limit=schemas.actions.ConcurrencyLimitV2Update(
+                    limit=concurrency_limit.concurrency_limit
+                ),
+            )
+            model = existing
+            model.limit = concurrency_limit.concurrency_limit
+        else:
+            model = await cl_v2_models.create_concurrency_limit(
+                session=session,
+                concurrency_limit=schemas.core.ConcurrencyLimitV2(
+                    name=v2_name,
+                    limit=concurrency_limit.concurrency_limit,
+                    active=True,
+                ),
+            )
+            try:
+                from prefect.settings.context import get_current_settings
 
-    return model
+                storage_mod = get_current_settings().server.concurrency.lease_storage
+                if storage_mod.endswith(".filesystem"):
+                    response.status_code = status.HTTP_201_CREATED
+                else:
+                    response.status_code = status.HTTP_200_OK
```

Comment:
This logic seems wrong. I think it's supposed to return 201 if a limit was created and 200 if it was updated, which shouldn't depend on the lease storage.